### ISSUE TITLE
Roadmap to Token Classification Tasks

### ISSUE THREAD

**Issue by shivance** [keras-team/keras-nlp#927]
# This roadmap needs inputs, edits and approval from @maintainers

In favor of #624 #754 

Thanks to @abuelnasr0 for ringing the bell. I think the work @TheAthleticCoder is doing for #741 and #841 can be reused in Token classification tasks as well as Span Labeling tasks. I'll try to put a roadmap on the same, clearly.

- keras_nlp uses tf_text as dependency for [WordPieceTokenizer](https://www.tensorflow.org/text/api_docs/python/text/FastWordpieceTokenizer) and [SentencePieceTokenizer](https://www.tensorflow.org/text/api_docs/python/text/SentencepieceTokenizer). Both of these tokenizers have `tokenize_with_offsets` (same as `split_with_offsets` ) methods, which should make the task easy for @TheAthleticCoder ( ? )

- Turns out that this method paves way for implementation of word to token mapping, [#754(comment_1)](https://github.com/keras-team/keras-nlp/issues/754#issuecomment-1435959495) the word_ids() method. We could do so by simply mapping each token to index of the sublist in the tokenized_with_offsets output ( I would put pseudo code if asked by maintainers )

- This essentially useful in general (apart from being backbone of Token Classification Task which demands mapping each label to each token, which then depends on the parent words where token came from). As @mattdangerw already mentioned in [#754(comment_2) ](https://github.com/keras-team/keras-nlp/issues/754#issuecomment-1441192086) that we will want to be able to run all our preprocessing inside a TF graph, indeed referring to #624 

   ![image](https://user-images.githubusercontent.com/51750587/227753906-491aac88-87cb-49ef-9417-a5983f7e5fd9.png)
- tf-text's [TokenizerWithOffsets](https://www.tensorflow.org/text/api_docs/python/text/TokenizerWithOffsets) can be referred to implement the functionality for BytePairTokenizer
- Once we are done with this, I think that would be the best time to ship #754. as it would be more generic and demonstrate the API of KerasNLP further.

- [ ] Adding `tokenize_with_offset` method in `WordPieceTokenizer` and `SentencePieceTokenizer`
- [ ] Adding `word_ids()` method in `WordPieceTokenizer` and `SentencePieceTokenizer`
- [ ] Adding Token Classification Keras-io example with the newly created functionalities
- [ ] Adding XXTokenClassification Tasks #624 
- [ ] Adding XXSpanLabeler (still needs some scoping) #625  
- [ ] Scoping tokenization with offsets for `BytePairTokenizer`

If this roadmap is approved, and as I and @TheAthleticCoder have been actively working on Token Classification and Squad Task respectively, it would be best that we collaborate to deliver these 2 features.

PS: This issue needs *scoping* and *enhancement* labels.

**Comment by abuelnasr0** [comment]
Thanks. I will create a PR for `tokenize_with_offset()` method for both `WordPieceTokenizer` and `SentencePieceTokenizer`, when I get the approval.

**Comment by TheAthleticCoder** [comment]
Yes, Sounds good @shivance!

**Comment by shivance** [comment]
> If this roadmap is approved, and as I and @TheAthleticCoder have been actively working on Token Classification and Squad Task respectively, it would be best to collaborate to deliver these 2 features.

@abuelnasr0 [#926/comment)](https://github.com/keras-team/keras-nlp/issues/926#issuecomment-1483907839), We thank you for your pointer, but I'm afraid that you can not claim the issues that I and @TheAthleticCoder have been working for a while now.

**Comment by abuelnasr0** [comment]
@shivance I will not take your issues. I will take my issue. I searched to get a method that can be added to tokenizers to return offset mapping. I will add offset mapping method and you can continue with your tasks using this method. may you just appreciate that I spent a lot of time exploring and searching for an issue to contribute in. I came up with this solution and I should commit it.

**Comment by shivance** [comment]
@TheAthleticCoder do you think that this infra can be useful for #625 ?

**Comment by shivance** [comment]
> @shivance I will not take your issues. I will take my issue. I searched to get a method that can be added to tokenizers to return offset mapping. I will add offset mapping method and you can continue with your tasks using this method. may you just appreciate that I spent a lot of time exploring and searching for an issue to contribute in. I came up with this solution and I should commit it.

Heyy @abuelnasr0 I already did appreciate with due credit to your finding, I think same conversation as #926 and some misunderstanding, Let's wait for @mattdangerw 's approval first? Then maybe we could break entire roadmap into small issues, this would be best to avoid conflict between contributors or stepping over each other's work.  And as you claim that you proposed addition of method `tokenize_with_offsets()` , then we'll let you work on that particular issue. Sounds good?

**Comment by abuelnasr0** [comment]
@shivance and that all I wanted only `tokenize_with_offsets()`. no more than that. 
thank you.

**Comment by shivance** [comment]
> @shivance and that all I wanted only `tokenize_with_offsets()`. no more than that. thank you.

Sure mate! Contribution & collaboration are heartly welcome in Open Source, but conflicts are not.

**Comment by mattdangerw** [comment]
Just a quick comment here.

I think we can stick the plan of starting with a keras.io example (tracked here https://github.com/keras-team/keras-nlp/issues/754), and using that to inform a lower level API.

I think `tokenize_with_offsets` is a good bet from brining "graph" support for the operations we need here, but I am still unsure if that is going to be the most user friendly function to expose on the tokenizer.

For now we can just get a guide working with the best way to do this today. Not that, while a little gross, the offsets API can be accessed today via `bert_tokenizer._fast_word_piece.tokenize_with_offsets`.

When we get a clear read on what API additions we would like here, we can open up more contributor issues for this! Thanks for the patience everyone!

**Comment by shivance** [comment]
Sure @mattdangerw, I would follow up on the corresponding PR on Keras-io !

**Comment by abuelnasr0** [comment]
`bert_tokenizer._fast_word_piece.tokenize_with_offsets` will not work because `no_pretokenization` was set to `True` when `_fast_word_piece` was instantiated and it will treat the whole sentence as one token. 
the solution is to create a new instance of `tf_text.FastWordpieceTokenizer` with `no_pretokenization` set to `False` and that will pretokenize on(white space, punctuation, chainese chars) but I think we can't create a new instance of `tf_text.FastWordpieceTokenizer`.
if you create a function like `pretokenize()` [here](https://github.com/keras-team/keras-nlp/blob/1795d1081183a41bf3de877f771cbee3be98749c/keras_nlp/tokenizers/word_piece_tokenizer.py#L98)  this function will convert the whole sentence to an array of words. and passing this to `bert_tokenizer._fast_word_piece.tokenize_with_offsets` will treat each word as a standalone sentence and will return only the start and end of each word independently however this can be solved, there's another crucial thing the whitespaces you removed in  `pretokenize()` is not counted in measuring start and end. if you can get up with a way to keep track of the white spaces that has been deleted and return them back to their position it will be good.

on the other hand `tf_text.SentencepieceTokenizer` is good to go.

that's is my experience with `tf_text.FastWordpieceTokenizer` and I think that will help you and shorten the way.
